File: python/bot/filters.py
"""
Custom filters for the Telegram Bot.
Filters for admin users and trusted (active) users.
"""
from telegram import Update
from telegram.ext import filters, BaseFilter

from config import Config
from db import database


class AdminFilter(BaseFilter):
    """Filter messages from the admin user only."""

    def filter(self, update: Update) -> bool:
        if not update.effective_user:
            return False
        return update.effective_user.id == Config.ADMIN_ID


class TrustedUserFilter(BaseFilter):
    """Filter messages from trusted (active) users, including admin."""

    async def filter_async(self, update: Update) -> bool:
        if not update.effective_user:
            return False
        # Admin is always trusted
        if update.effective_user.id == Config.ADMIN_ID:
            return True
        user = await database.get_user_by_tg_id(update.effective_user.id)
        return user is not None and user.get("is_active", 0) == 1


# Singleton instances
admin_filter = AdminFilter()
trusted_user_filter = TrustedUserFilter()

File: nodejs/src/bot/filters.ts
import type { Context } from "grammy";
import { config } from "../config.js";
import { getUserByTgId } from "../db/database.js";

/**
 * 檢查是否為管理員
 */
export function isAdmin(ctx: Context): boolean {
  return ctx.from?.id === config.ADMIN_ID;
}

/**
 * 檢查是否為信任使用者（DB 中存在且 is_active，管理員永遠是信任的）
 */
export function isTrustedUser(ctx: Context): boolean {
  if (isAdmin(ctx)) return true;

  const tgId = ctx.from?.id;
  if (!tgId) return false;

  const user = getUserByTgId(tgId);
  return user !== undefined && Number(user.is_active) === 1;
}
